src/potentials.cpp

```diff
@@ -504,92 +504,171 @@ void FunctorPotential::from_json(const json &j) {
+}
 
-                assert(rmin2 >= 0);
+void TabulatedPotential::from_json(const json &js) {
+    FunctorPotential::from_json(js);
+    if (!isotropic) {
+        throw std::runtime_error("Cannot spline anisotropic potentials");
+    }
+    spline.setTolerance(js.value("utol", 1e-3), js.value("ftol", 1e-2));
+    hardsphere_repulsion = js.value("hardsphere", false);
+    double energy_at_rmin = js.value("u_at_rmin", 20);
+    double energy_at_rmax = js.value("u_at_rmax", 1e-6);
 
-                while (rmax2 >= dr) {
-                    double u = std::fabs(this->umatrix(i, k)(a, b, rmax2, {0, 0, sqrt(rmax2)}));
-                    if (u > u_at_rmax)
-                        rmax2 = rmax2 + dr;
-                    else
-                        break;
-                }
+    faunus_logger->trace("Pair potential spline tolerance = {} kT", js.value("utol", 1e-5));
 
-                assert(rmin2 < rmax2);
-
-                KnotData knotdata = spline.generate(
-                    [&](double r2) {
-                        return this->umatrix(i, k)(a, b, r2, {0, 0, 0});
-                    },
-                    rmin2, rmax2);
-
-                // assert if potential is negative for r<rmin
-                if (spline.eval(knotdata, knotdata.rmin2 + dr) < 0) {
-                    assert(hardsphere == false && "`hardsphere` is set, but potential is negative for r<rmin");
-                    knotdata.isNegativeBelowRmin = true;
-                } else
-                    matrix_of_knots.set(i, k, knotdata);
-
-                faunus_logger->debug("Potential for {}-{} splined with {} knot(s)", atoms[i].name, atoms[k].name,
-                                     knotdata.numKnots());
-
-                if (j.value("to_disk", false)) {
-                    std::ofstream f(atoms[i].name + "-" + atoms[k].name + "_tabulated.dat"); // output file
-                    f << "# r splined exact\n";
-                    Point r = {dr, 0, 0}; // variable distance vector between particle a and b
-                    for (; r.x() < sqrt(rmax2); r.x() += dr)
-                        f << r.x() << " " << operator()(a, b, r.x() * r.x(), r) << " "
-                          << this->umatrix(i, k)(a, b, r.x() * r.x(), r) << "\n";
+    for (size_t i = 0; i < Faunus::atoms.size(); ++i) { // loop over atom types
+        for (size_t j = 0; j <= i; ++j) {               // and build matrix of spline data (knots) for each pair
+            if (atoms[i].implicit || atoms[j].implicit) {
+                continue;
+            }
+            double rmin = 0.5 * (Faunus::atoms[i].sigma + Faunus::atoms[j].sigma);
+            double rmax = js.value("rmax", rmin * 10);
+            if (auto it = js.find("cutoff_g2g"); it != js.end()) {
+                if (it->is_number()) {
+                    rmax = it->get<double>();
+                } else if (it->is_object()) {
+                    rmax = it->at("default").get<double>();
                 }
             }
```

Comment:
yes, it's the initial guess and it could be improved by also searching in the other direction (less knots). I don't want to touch it before talking to @gitesei who may have a use-case for the (currently undocumented) `group2group` threshold.